File: backend/app/services/clustering.py
"""
Clustering service - Topic detection and organization using OpenRouter LLM.
"""
import json
import logging
from pathlib import Path
from typing import List, Dict, Any
from openai import AsyncOpenAI

logger = logging.getLogger(__name__)

# OpenRouter configuration
OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1"
CLUSTERING_MODEL = "tngtech/deepseek-r1t2-chimera:free"

# Load prompt from markdown file
_prompt_path = Path(__file__).parent.parent / "prompt" / "clustering.md"
CLUSTERING_PROMPT = _prompt_path.read_text(encoding="utf-8")


class ClusteringService:
    """Handles topic clustering using OpenRouter LLM."""
    
    async def analyze_and_cluster(
        self, 
        texts: List[Dict[str, str]], 
        api_key: str
    ) -> Dict[str, Any]:
        """
        Analyze extracted texts and create topic clusters.
        
        Args:
            texts: List of dicts with 'filename' and 'content' keys
            api_key: OpenRouter API key
            
        Returns:
            Clustering result with identified topics
        """
        # Configure OpenRouter client
        client = AsyncOpenAI(
            base_url=OPENROUTER_BASE_URL,
            api_key=api_key,
        )
        
        # Combine texts with source markers
        combined_text = self._prepare_text(texts)
        
        # Generate clustering - append input to prompt template
        prompt = f"{CLUSTERING_PROMPT}\n\n## Input Materials\n\n{combined_text[:50000]}"
        
        try:
            response = await client.chat.completions.create(
                model=CLUSTERING_MODEL,
                messages=[{"role": "user", "content": prompt}],
                temperature=0.3,
                max_tokens=4096,
            )
            
            # Parse JSON response
            result = self._parse_response(response.choices[0].message.content)
            return result
            
        except Exception as e:
            logger.error(f"Clustering failed: {e}")
            raise
    
    def _prepare_text(self, texts: List[Dict[str, str]]) -> str:
        """Prepare combined text with source markers."""
        parts = []
        for item in texts:
            parts.append(f"\n=== SOURCE: {item['filename']} ===\n")
            parts.append(item['content'])
        return "\n".join(parts)
    
    def _parse_response(self, response_text: str) -> Dict[str, Any]:
        """Parse and validate Gemini response."""
        # Clean up response - extract JSON
        text = response_text.strip()
        
        # Find JSON block
        if "```json" in text:
            start = text.find("```json") + 7
            end = text.find("```", start)
            text = text[start:end]
        elif "```" in text:
            start = text.find("```") + 3
            end = text.find("```", start)
            text = text[start:end]
        
        try:
            result = json.loads(text)
            
            # Validate structure
            if "clusters" not in result:
                raise ValueError("Missing 'clusters' in response")
            
            return result
            
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse clustering response: {e}")
            # Return fallback structure
            return {
                "clusters": [{
                    "id": "cluster_1",
                    "title": "All Content",
                    "keywords": [],
                    "source_mapping": [],
                    "estimated_word_count": 0,
                    "summary": "Could not automatically cluster content"
                }],
                "total_clusters": 1,
                "processing_notes": "Fallback due to parsing error"
            }


# Singleton instance
clustering_service = ClusteringService()

File: frontend/src/lib/clustering.ts
/**
 * Clustering Service - Analyze documents and create topic clusters
 */
import { getLLMClient, PROMPTS } from './llm';
import { getSettings, Document } from './db';

// ===== Types =====

export interface ClusterSourceMapping {
  source: string;
  slides?: number[];
}

export interface ClusterResult {
  id: string;
  title: string;
  keywords: string[];
  sourceMapping: ClusterSourceMapping[];
  summary: string;
  estimatedWordCount: number;
  uniqueConcepts: string[];
}

export interface ClusteringResponse {
  clusters: ClusterResult[];
  totalClusters: number;
  uniquenessVerification: string;
}

// ===== Clustering Functions =====

/**
 * Analyze documents and generate topic clusters
 */
export async function analyzeAndCluster(documents: Document[]): Promise<ClusteringResponse> {
  if (!documents.length) {
    throw new Error('No documents to analyze');
  }

  const client = await getLLMClient();
  const settings = await getSettings();

  // Build source content with markers
  const sourceContent = documents
    .map(doc => `=== SOURCE: ${doc.filename} ===\n${doc.content}`)
    .join('\n\n');

  const prompt = `${PROMPTS.clustering}

---

## Documents to Analyze

${sourceContent.slice(0, 50000)}
`;

  try {
    const response = await client.chat.completions.create({
      model: settings.model,
      messages: [{ role: 'user', content: prompt }],
      temperature: 0.3, // Lower temperature for more consistent JSON
      max_tokens: 4096,
    });

    const content = response.choices[0]?.message?.content || '';
    return parseClusterResponse(content);
  } catch (error) {
    console.error('Clustering analysis failed:', error);
    throw error;
  }
}

/**
 * Parse clustering response JSON
 */
function parseClusterResponse(content: string): ClusteringResponse {
  // Extract JSON from response (might be wrapped in markdown code block)
  let jsonStr = content.trim();
  
  // Remove markdown code blocks if present
  const jsonMatch = jsonStr.match(/```(?:json)?\s*([\s\S]*?)```/);
  if (jsonMatch) {
    jsonStr = jsonMatch[1].trim();
  }

  try {
    const parsed = JSON.parse(jsonStr);
    
    // Validate and transform response
    const clusters: ClusterResult[] = (parsed.clusters || []).map((c: any, index: number) => ({
      id: c.id || String(index + 1),
      title: c.title || `Cluster ${index + 1}`,
      keywords: Array.isArray(c.keywords) ? c.keywords : [],
      sourceMapping: Array.isArray(c.source_mapping) 
        ? c.source_mapping.map((s: any) => ({
            source: s.source || '',
            slides: Array.isArray(s.slides) ? s.slides : [],
          }))
        : [],
      summary: c.summary || '',
      estimatedWordCount: c.estimated_word_count || 0,
      uniqueConcepts: Array.isArray(c.unique_concepts) ? c.unique_concepts : [],
    }));

    return {
      clusters,
      totalClusters: parsed.total_clusters || clusters.length,
      uniquenessVerification: parsed.uniqueness_verification || '',
    };
  } catch (error) {
    console.error('Failed to parse clustering response:', error);
    console.error('Raw content:', content);
    
    // Return empty result on parse failure
    throw new Error('Failed to parse clustering response. Please try again.');
  }
}

/**
 * Get combined content for a cluster from documents
 */
export function getClusterContent(
  cluster: ClusterResult,
  documents: Document[]
): string {
  const relevantDocs = documents.filter(doc => 
    cluster.sourceMapping.some(mapping => 
      doc.filename.includes(mapping.source) || mapping.source.includes(doc.filename)
    )
  );

  if (relevantDocs.length === 0) {
    // Fallback: use all documents if no mapping match
    return documents.map(d => d.content).join('\n\n---\n\n');
  }

  return relevantDocs.map(d => d.content).join('\n\n---\n\n');
}

/**
 * Create manual cluster (when user wants to skip auto-clustering)
 */
export function createManualCluster(
  title: string,
  documents: Document[]
): ClusterResult {
  return {
    id: crypto.randomUUID(),
    title,
    keywords: [],
    sourceMapping: documents.map(d => ({
      source: d.filename,
      slides: [],
    })),
    summary: `Manual cluster containing: ${documents.map(d => d.filename).join(', ')}`,
    estimatedWordCount: documents.reduce((sum, d) => sum + d.content.split(/\s+/).length, 0),
    uniqueConcepts: [],
  };
}

/**
 * Merge multiple clusters into one
 */
export function mergeClusters(
  clusters: ClusterResult[],
  newTitle: string
): ClusterResult {
  return {
    id: crypto.randomUUID(),
    title: newTitle,
    keywords: [...new Set(clusters.flatMap(c => c.keywords))],
    sourceMapping: clusters.flatMap(c => c.sourceMapping),
    summary: clusters.map(c => c.summary).join(' | '),
    estimatedWordCount: clusters.reduce((sum, c) => sum + c.estimatedWordCount, 0),
    uniqueConcepts: [...new Set(clusters.flatMap(c => c.uniqueConcepts))],
  };
}
